## Ejercicios pandas 1

In [2]:
import pandas as pd

### Ejercicio 1:
- Lee el archivo **"Canada.xlsx"** y guardalo en la variable **`df`**

In [3]:
df = pd.read_excel("../Actividades/Canada.xlsx")

### Ejercicio 2:
- Mostrar el tamaño de DataFrame.

In [4]:
tamaño = {"Filas" : df.shape[0],
          "Columnas" : df.shape[1]}

tamaño

{'Filas': 195, 'Columnas': 41}

### Ejercicio 3:
- Mostrar las primeras 3 filas.

In [5]:
df.head(3)

Tipo         Pais  Area Continente  Region Nombre de region  \
0  Immigrants  Afghanistan   935       Asia    5501    Southern Asia   
1  Immigrants      Albania   908     Europe     925  Southern Europe   
2  Immigrants      Algeria   903     Africa     912  Northern Africa   

       Tipo de region  1980  1981  1982  ...  2004  2005  2006  2007  2008  \
0  Developing regions    16    39    39  ...  2978  3436  3009  2652  2111   
1   Developed regions     1     0     0  ...  1450  1223   856   702   560   
2  Developing regions    80    67    71  ...  3616  3626  4807  3623  4005   

   2009  2010  2011  2012  2013  
0  1746  1758  2203  2635  2004  
1   716   561   539   620   603  
2  5393  4752  4325  3774  4331  

[3 rows x 41 columns]

### Ejercicio 4:
- Mostrar las últimas 10 filas.

In [6]:
df.tail(10)

Tipo                                Pais  Area  \
185  Immigrants            United States of America   905   
186  Immigrants                             Uruguay   904   
187  Immigrants                          Uzbekistan   935   
188  Immigrants                             Vanuatu   909   
189  Immigrants  Venezuela (Bolivarian Republic of)   904   
190  Immigrants                            Viet Nam   935   
191  Immigrants                      Western Sahara   903   
192  Immigrants                               Yemen   935   
193  Immigrants                              Zambia   903   
194  Immigrants                            Zimbabwe   903   

                          Continente  Region    Nombre de region  \
185                 Northern America     905    Northern America   
186  Latin America and the Caribbean     931       South America   
187                             Asia    5500        Central Asia   
188                          Oceania     928           Melanesia   
189  Latin America and the Caribbean     931       South America   
190                             Asia     920  South-Eastern Asia   
191                           Africa     912     Northern Africa   
192                             Asia     922        Western Asia   
193                           Africa     910      Eastern Africa   
194                           Africa     910      Eastern Africa   

         Tipo de region  1980   1981  1982  ...  2004  2005  2006  2007  \
185   Developed regions  9378  10030  9074  ...  6990  8394  9613  9463   
186  Developing regions   128    132   146  ...   130   217   175   147   
187  Developing regions     0      0     0  ...   175   330   262   284   
188  Developing regions     0      0     0  ...     1     1     1     0   
189  Developing regions   103    117   174  ...  1224  1211  1192  1335   
190  Developing regions  1191   1829  2162  ...  1816  1852  3153  2574   
191  Developing regions     0      0     0  ...     0     0     1     0   
192  Developing regions     1      2     1  ...   124   161   140   122   
193  Developing regions    11     17    11  ...    56    91    77    71   
194  Developing regions    72    114   102  ...  1450   615   454   663   

      2008  2009  2010  2011  2012  2013  
185  10190  8995  8142  7676  7891  8501  
186    160    99    93    81    47    58  
187    215   288   289   162   235   167  
188      0     1     0     0     0     0  
189   1239  1353   998  1452  1373  1022  
190   1784  2171  1942  1723  1731  2112  
191      0     0     0     0     0     0  
192    133   128   211   160   174   217  
193     64    60   102    69    46    59  
194    611   508   494   434   437   407  

[10 rows x 41 columns]

### Ejercicio 5:
- Mostrar las estadisticas del DataFrame.

In [7]:
df.describe()

Area       Region          1980          1981          1982  \
count  195.000000   195.000000    195.000000    195.000000    195.000000   
mean   912.764103  1249.015385    508.394872    566.989744    534.723077   
std     13.082835  1185.526885   1949.588546   2152.643752   1866.997511   
min    903.000000   905.000000      0.000000      0.000000      0.000000   
25%    903.000000   914.000000      0.000000      0.000000      0.000000   
50%    908.000000   922.000000     13.000000     10.000000     11.000000   
75%    922.000000   925.500000    251.500000    295.500000    275.000000   
max    935.000000  5501.000000  22045.000000  24796.000000  20620.000000   

               1983          1984         1985         1986          1987  \
count    195.000000    195.000000   195.000000   195.000000    195.000000   
mean     387.435897    376.497436   358.861538   441.271795    691.133333   
std     1204.333597   1198.246371  1079.309600  1225.576630   2109.205607   
min        0.000000      0.000000     0.000000     0.000000      0.000000   
25%        0.000000      0.000000     0.000000     0.500000      0.500000   
50%       12.000000     13.000000    17.000000    18.000000     26.000000   
75%      173.000000    181.000000   197.000000   254.000000    434.000000   
max    10015.000000  10170.000000  9564.000000  9470.000000  21337.000000   

       ...          2004          2005          2006          2007  \
count  ...    195.000000    195.000000    195.000000    195.000000   
mean   ...   1190.169231   1320.292308   1266.958974   1191.820513   
std    ...   3710.505369   4425.957828   3926.717747   3443.542409   
min    ...      0.000000      0.000000      0.000000      0.000000   
25%    ...     19.000000     28.500000     25.000000     31.000000   
50%    ...    191.000000    210.000000    218.000000    198.000000   
75%    ...    756.500000    832.000000    842.000000    899.000000   
max    ...  36619.000000  42584.000000  33848.000000  28742.000000   

               2008          2009          2010          2011          2012  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean    1246.394872   1275.733333   1420.287179   1262.533333   1313.958974   
std     3694.573544   3829.630424   4462.946328   4030.084313   4247.555161   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%       31.000000     36.000000     40.500000     37.500000     42.500000   
50%      205.000000    214.000000    211.000000    179.000000    233.000000   
75%      934.500000    888.000000    932.000000    772.000000    783.000000   
max    30037.000000  29622.000000  38617.000000  36765.000000  34315.000000   

               2013  
count    195.000000  
mean    1320.702564  
std     4237.951988  
min        0.000000  
25%       45.000000  
50%      213.000000  
75%      796.000000  
max    34129.000000  

[8 rows x 36 columns]

In [8]:
#mostrar la estadistica de una columna solo:
df[1991].describe()

count      195.000000
mean      1064.148718
std       2952.093731
min          0.000000
25%          3.000000
50%         51.000000
75%        657.500000
max      23380.000000
Name: 1991, dtype: float64

In [9]:
#estadisticas de un rango de columnas:

df.loc[:, 1991:2000].describe()

1991          1992          1993          1994          1995  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean    1064.148718   1136.856410   1138.712821    993.153846    962.625641   
std     2952.093731   3330.083742   3495.220063   3613.336444   3091.492343   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        3.000000      6.500000     11.500000      9.500000     10.500000   
50%       51.000000     74.000000     85.000000     76.000000     91.000000   
75%      657.500000    655.000000    722.500000    545.000000    550.500000   
max    23380.000000  34123.000000  33720.000000  39231.000000  30145.000000   

               1996          1997          1998          1999          2000  
count    195.000000    195.000000    195.000000    195.000000    195.000000  
mean    1026.076923    989.153846    824.241026    922.143590   1111.343590  
std     3321.045004   3070.761447   2385.943695   2887.632585   3664.042361  
min        0.000000      0.000000      0.000000      0.000000      0.000000  
25%       14.500000     19.500000     15.000000     16.000000     16.000000  
50%      118.000000    114.000000    106.000000    116.000000    138.000000  
75%      603.500000    612.500000    535.500000    548.500000    659.000000  
max    29322.000000  22965.000000  21049.000000  30069.000000  35529.000000

### Ejercicio 6:
- Mostrar la información del DataFrame.

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Data columns (total 41 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Tipo              195 non-null    object
 1   Pais              195 non-null    object
 2   Area              195 non-null    int64 
 3   Continente        195 non-null    object
 4   Region            195 non-null    int64 
 5   Nombre de region  195 non-null    object
 6   Tipo de region    195 non-null    object
 7   1980              195 non-null    int64 
 8   1981              195 non-null    int64 
 9   1982              195 non-null    int64 
 10  1983              195 non-null    int64 
 11  1984              195 non-null    int64 
 12  1985              195 non-null    int64 
 13  1986              195 non-null    int64 
 14  1987              195 non-null    int64 
 15  1988              195 non-null    int64 
 16  1989              195 non-null    int64 
 17  1990            

### Ejercicio 7:
- Mostar las columnas del DataFrame en una lista.

In [11]:
lista_de_columnas = df.columns.tolist()
print(lista_de_columnas)


['Tipo', 'Pais', 'Area', 'Continente', 'Region', 'Nombre de region', 'Tipo de region', 1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013]


In [12]:
#crear una lista de columnas determinadas: Aquí necesitaremos entrar en las columnas por su indice no por su nombre:
columnas_rango = df.iloc[:, 18:28].columns.tolist()
print(columnas_rango)

[1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000]


In [13]:
#crear una lista de columnas, escribiendo sus nombres:
columnas_1991_2000 = df[[1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000]].columns.tolist()
print(columnas_1991_2000)

[1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000]


### Ejercicio 8:
- Eliminar las columnas: **`Tipo`**, **`Area`** y **`Region`**.

In [14]:
# en el drop, rename, etc: axis = 1 significa columnas y axis=0 significa filas
#en los demás metodos como max, min, median, etc: axis = 1 significa filas y axis = 0 significa columna

Se está eliminando una columna, lo que implica una operación a lo largo del eje horizontal (a través de las columnas). Por lo tanto, la orientación es horizontal y axis=1 es apropiado en este contexto.

La clave aquí es que, a pesar de que los datos están dispuestos verticalmente en una columna, desde la perspectiva de la operación de eliminación, se está tratando con las columnas como una entidad colectiva. Por eso, axis=1 indica que la operación se realiza en la dirección de las columnas.

In [15]:
#HAY QUE PONER AXIS = 1 PARA QUE ELIMINE LA COLUMNA HACIA ABAJO, dentro del parentesis al lado del nombre de la columna
#Si omites el argumento axis=1 al utilizar el método drop en un DataFrame, pandas podría generar un error porque necesita saber si estás tratando de eliminar una columna o una fila.
df.drop("Tipo", axis = 1, inplace = True)  #inplace = true dentro del parentesis del metodo

In [16]:
#eliminar varias columnas a la vez, las columnas tienen que estar en forma de lista:
df.drop(["Area", "Region"], axis = 1, inplace = True)

### Ejercicio 9:
- Obtener los valores únicos de la columna **`Continente`** en una lista.

In [17]:
valores_unicos = df["Continente"].unique().tolist()

In [18]:
valores_unicos

['Asia',
 'Europe',
 'Africa',
 'Oceania',
 'Latin America and the Caribbean',
 'Northern America']

### Ejercicio 10:
- Obtener el valor minimo del año 2000.

In [19]:
maximo_columna_2000 = df[2000].max()

In [20]:
maximo_columna_2000

35529

### Ejercicio 11:
- Obtener los valores máximos de 1991-2000.

In [21]:
#crear una lista de columnas, escribiendo sus nombres:
columnas_1991_2000 = df[[1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000]].columns.tolist()
print(columnas_1991_2000)

[1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000]


In [22]:
df[columnas_1991_2000].max()

1991    23380
1992    34123
1993    33720
1994    39231
1995    30145
1996    29322
1997    22965
1998    21049
1999    30069
2000    35529
dtype: int64

Para encontrar los valores máximos en un rango específico de columnas (en este caso, de la columna '1991' a la columna '2000') en un DataFrame de pandas, puedes usar el método max() junto con el argumento axis=1 para calcular el máximo a lo largo de las columnas:

In [23]:
maximos_decadafilas_entre_lascolumnas1991_2000 = df.loc[:, 1991:2000].max(axis=1) #EJE HORIZONTAL

In [24]:
maximos_decadafilas_entre_lascolumnas1991_2000

0      3326
1      1816
2      2867
3         0
4         2
       ... 
190    6547
191       1
192     144
193      73
194      98
Length: 195, dtype: int64

In [25]:
#LO MISMO QUE EN LA CELDA 69
maximos_1991_2000 = df.loc[:, 1991:2000].max(axis=0)

In [26]:
maximos_1991_2000

1991    23380
1992    34123
1993    33720
1994    39231
1995    30145
1996    29322
1997    22965
1998    21049
1999    30069
2000    35529
dtype: int64

In [27]:
#mostrar en qué indice está el valor tal en la columna 1991

df[df[1991] == 23380].index[0]

183

In [28]:
#Si deseas mostrar todo el contenido de un DataFrame sin que se trunquen las filas, puedes configurar la opción de visualización de pandas para que no limite el número de filas que se muestran. Esto se hace usando:
pd.set_option('display.max_rows', None)

In [29]:
#Después de ejecutar este código, cuando imprimas el DataFrame, deberías ver todas las filas sin truncar. Si deseas volver a la configuración predeterminada, puedes hacerlo usando:
pd.reset_option('display.max_rows')

In [30]:
#mostrar los valores de la columna 1991 en un rango:
df.loc[:, 1991]

0      1378
1        21
2       872
3         0
4         0
       ... 
190    5870
191       1
192      18
193      69
194      94
Name: 1991, Length: 195, dtype: int64

In [31]:
df

Pais Continente    Nombre de region      Tipo de region  1980  \
0       Afghanistan       Asia       Southern Asia  Developing regions    16   
1           Albania     Europe     Southern Europe   Developed regions     1   
2           Algeria     Africa     Northern Africa  Developing regions    80   
3    American Samoa    Oceania           Polynesia  Developing regions     0   
4           Andorra     Europe     Southern Europe   Developed regions     0   
..              ...        ...                 ...                 ...   ...   
190        Viet Nam       Asia  South-Eastern Asia  Developing regions  1191   
191  Western Sahara     Africa     Northern Africa  Developing regions     0   
192           Yemen       Asia        Western Asia  Developing regions     1   
193          Zambia     Africa      Eastern Africa  Developing regions    11   
194        Zimbabwe     Africa      Eastern Africa  Developing regions    72   

     1981  1982  1983  1984  1985  ...  2004  2005  2006  2007  2008  2009  \
0      39    39    47    71   340  ...  2978  3436  3009  2652  2111  1746   
1       0     0     0     0     0  ...  1450  1223   856   702   560   716   
2      67    71    69    63    44  ...  3616  3626  4807  3623  4005  5393   
3       1     0     0     0     0  ...     0     0     1     0     0     0   
4       0     0     0     0     0  ...     0     0     1     1     0     0   
..    ...   ...   ...   ...   ...  ...   ...   ...   ...   ...   ...   ...   
190  1829  2162  3404  7583  5907  ...  1816  1852  3153  2574  1784  2171   
191     0     0     0     0     0  ...     0     0     1     0     0     0   
192     2     1     6     0    18  ...   124   161   140   122   133   128   
193    17    11     7    16     9  ...    56    91    77    71    64    60   
194   114   102    44    32    29  ...  1450   615   454   663   611   508   

     2010  2011  2012  2013  
0    1758  2203  2635  2004  
1     561   539   620   603  
2    4752  4325  3774  4331  
3       0     0     0     0  
4       0     0     1     1  
..    ...   ...   ...   ...  
190  1942  1723  1731  2112  
191     0     0     0     0  
192   211   160   174   217  
193   102    69    46    59  
194   494   434   437   407  

[195 rows x 38 columns]

### Ejercicio 12:
- Obtener una lista con la media de migración de todos los años.

In [32]:
#obtener una lista con los nombres solamente de los paises dentro de la columna pais:

In [33]:
paises_migratorios = df["Pais"].tolist() #Si pusiera valores unique() me arriesgo a que haya filas con paises repetidos y no salgan en la lista
len(paises_migratorios)

195

In [34]:
media_cadafila = maximos_1991_2000 = df.loc[:, 1980:2013].median(axis=1).tolist()
len(media_cadafila)

195

In [35]:
diccionario_migracion = {}
for pais, media in zip(paises_migratorios, media_cadafila):
    diccionario_migracion[pais] = media

In [36]:
diccionario_migracion

{'Afghanistan': 1752.0,
 'Albania': 210.0,
 'Algeria': 1948.0,
 'American Samoa': 0.0,
 'Andorra': 0.0,
 'Angola': 32.5,
 'Antigua and Barbuda': 26.0,
 'Argentina': 443.0,
 'Armenia': 88.0,
 'Australia': 700.5,
 'Austria': 131.5,
 'Azerbaijan': 32.0,
 'Bahamas': 21.0,
 'Bahrain': 12.0,
 'Bangladesh': 2084.0,
 'Barbados': 164.5,
 'Belarus': 201.5,
 'Belgium': 307.0,
 'Belize': 33.0,
 'Benin': 27.0,
 'Bhutan': 2.0,
 'Bolivia (Plurinational State of)': 73.5,
 'Bosnia and Herzegovina': 175.0,
 'Botswana': 4.0,
 'Brazil': 697.5,
 'Brunei Darussalam': 6.0,
 'Bulgaria': 709.5,
 'Burkina Faso': 20.5,
 'Burundi': 164.0,
 'Cabo Verde': 4.5,
 'Cambodia': 209.5,
 'Cameroon': 150.0,
 'Canada': 0.0,
 'Central African Republic': 6.0,
 'Chad': 28.0,
 'Chile': 399.0,
 'China': 19945.0,
 'China, Hong Kong Special Administrative Region': 12.0,
 'China, Macao Special Administrative Region': 0.0,
 'Colombia': 633.0,
 'Comoros': 3.0,
 'Congo': 20.0,
 'Costa Rica': 114.5,
 "Côte d'Ivoire": 84.5,
 'Croatia': 

### Ejercicio 13:
- Obtener el país con más migración en el año 2010.

In [37]:
#obtenemos el indice donde el valor en la columna 2010 es maximo:
indice = df[2010].idxmax()

In [38]:
#vemos a que pais corresponde ese indice, en la columna pais:
df.at[indice, "Pais"]

'Philippines'

In [39]:
#En una sola linea, con loc, accedo al lugar dado un indice y el nombre de la columna donde quiero acceder:
df.loc[df[2010].idxmax(), "Pais"]

'Philippines'

In [40]:
#Si tuviesemos varios indices que tienen el mismo maximo en la columna 2010 y quisieramos que muestre todos, y no solo el primero como anteriormente:
df.loc[df[2010] == df[2010].max(), 'Pais'].tolist()

['Philippines']

### Ejercicio 14
- Obtener el país con menor migración en el año 2005.

In [41]:
df.loc[df[2010].idxmin(), "Pais"]  #aqui te devuelve el primero que encuentre

'American Samoa'

In [42]:
#Si tuviesemos varios indices que tienen el mismo minimo en la columna 2010 y quisieramos que muestre todos, y no solo el primero como anteriormente:
df.loc[df[2010] == df[2010].min(), "Pais"].tolist()

['American Samoa',
 'Andorra',
 'Canada',
 'Kiribati',
 'Liechtenstein',
 'Marshall Islands',
 'Nauru',
 'New Caledonia',
 'Palau',
 'Tuvalu',
 'Vanuatu',
 'Western Sahara']

In [43]:
#voy a comprobarlo accediendo a traves del nombre del pais al valor en la columna 2010:
for nombre in df.loc[df[2010] == df[2010].min(), "Pais"].tolist():
    print(nombre, df.loc[df['Pais'] == nombre, 2010].values[0])

American Samoa 0
Andorra 0
Canada 0
Kiribati 0
Liechtenstein 0
Marshall Islands 0
Nauru 0
New Caledonia 0
Palau 0
Tuvalu 0
Vanuatu 0
Western Sahara 0


In [44]:
#voy a comprobarlo con mi diccionario de migracion que me da las medias:
diccionario_migracion["American Samoa"], diccionario_migracion["Andorra"], diccionario_migracion["Canada"]

(0.0, 0.0, 0.0)

### Ejercicio 15:
- Encuentra el año en que más migración hubo de España a Canada.

In [45]:
#encontrar la columna donde hay un maximo dando el valor en otra columna:
#loc[[accedemos al valor "Spain" en la columna "Pais"].loc[busca el indice con valor maximo de todas las filas, desde la columna 4 en adelante, horizontalmente]]
#.values[0]: Esto convierte el resultado en un arreglo de NumPy y luego obtiene el primer elemento. Esto es necesario porque el resultado es un objeto de pandas y queremos el valor real como una cadena.
df.loc[df['Pais'] == "Spain"].iloc[:, 4:].idxmax(axis=1).values[0]

2013

### Ejercicio 16:
- Encuentra el año en que menor migración hubo de España a Canada.

In [46]:
df.loc[df['Pais'] == "Spain"].iloc[:, 4:].idxmin(axis=1).values[0]

1995

### Ejercicio 17:
- Ordena en una lista los años de migración en España de menor a mayor de acuerdo a la información de cada año.

In [47]:
# Seleccionar las columnas de migración para España
columnas_spain = df.loc[df['Pais'] == 'Spain', 1980:2013]
#columnas_spain.squeeze(): squeeze() se utiliza para convertir un DataFrame con una sola columna en una Serie. En este caso, columnas_spain es un DataFrame con una sola columna (los datos de inmigración desde España en diferentes años). 
#Entonces, este paso convierte columnas_spain en una Serie:
lista_de_columnas = columnas_spain.squeeze().sort_values().index.tolist()
#.sort_values(): Esto ordena los valores en la Serie en orden ascendente, ya que accede a la fila Spain y a los datos de cada columna años en esa fila.
lista_de_columnas

[1995,
 2001,
 1999,
 1994,
 2002,
 1998,
 1997,
 2003,
 2000,
 1985,
 1996,
 1986,
 1990,
 2006,
 2005,
 1992,
 2004,
 1983,
 2007,
 1988,
 1984,
 1993,
 1991,
 1989,
 2008,
 2010,
 1987,
 2011,
 2009,
 1980,
 1982,
 2012,
 1981,
 2013]

### Ejercicio 18:
- Crea una lista con todos los países.

In [48]:
paises_migratorios = df["Pais"].tolist() #Si pusiera valores unique() me arriesgo a que haya filas con paises repetidos y no salgan en la lista
len(paises_migratorios)

195

In [49]:
df

Pais Continente    Nombre de region      Tipo de region  1980  \
0       Afghanistan       Asia       Southern Asia  Developing regions    16   
1           Albania     Europe     Southern Europe   Developed regions     1   
2           Algeria     Africa     Northern Africa  Developing regions    80   
3    American Samoa    Oceania           Polynesia  Developing regions     0   
4           Andorra     Europe     Southern Europe   Developed regions     0   
..              ...        ...                 ...                 ...   ...   
190        Viet Nam       Asia  South-Eastern Asia  Developing regions  1191   
191  Western Sahara     Africa     Northern Africa  Developing regions     0   
192           Yemen       Asia        Western Asia  Developing regions     1   
193          Zambia     Africa      Eastern Africa  Developing regions    11   
194        Zimbabwe     Africa      Eastern Africa  Developing regions    72   

     1981  1982  1983  1984  1985  ...  2004  2005  2006  2007  2008  2009  \
0      39    39    47    71   340  ...  2978  3436  3009  2652  2111  1746   
1       0     0     0     0     0  ...  1450  1223   856   702   560   716   
2      67    71    69    63    44  ...  3616  3626  4807  3623  4005  5393   
3       1     0     0     0     0  ...     0     0     1     0     0     0   
4       0     0     0     0     0  ...     0     0     1     1     0     0   
..    ...   ...   ...   ...   ...  ...   ...   ...   ...   ...   ...   ...   
190  1829  2162  3404  7583  5907  ...  1816  1852  3153  2574  1784  2171   
191     0     0     0     0     0  ...     0     0     1     0     0     0   
192     2     1     6     0    18  ...   124   161   140   122   133   128   
193    17    11     7    16     9  ...    56    91    77    71    64    60   
194   114   102    44    32    29  ...  1450   615   454   663   611   508   

     2010  2011  2012  2013  
0    1758  2203  2635  2004  
1     561   539   620   603  
2    4752  4325  3774  4331  
3       0     0     0     0  
4       0     0     1     1  
..    ...   ...   ...   ...  
190  1942  1723  1731  2112  
191     0     0     0     0  
192   211   160   174   217  
193   102    69    46    59  
194   494   434   437   407  

[195 rows x 38 columns]

### Ejercicio 19:
- Encuentra el número de paises por continente y agrega los nombres a un diccionario.
    - Ejemplo: **`{"Asía" : 20, "Africa" : 10, "Europa" : 5}`**

In [50]:
# metodo para contar la cantidad de elementos iguales que hay en una columna
# .value_counts()
# el metodo to_dict() directamente te convierte en un diccionario cuyas llaves serán los valores de la columna seleccionada y el valor será lo que hayas hecho con esa columna

In [51]:
diccionario_continentes = df["Continente"].value_counts().to_dict()
diccionario_continentes

{'Africa': 54,
 'Asia': 49,
 'Europe': 43,
 'Latin America and the Caribbean': 33,
 'Oceania': 14,
 'Northern America': 2}

### Ejercicio 20:
- Crea un DataFrame a partir del diccionario anterior.

In [52]:
# Crear el DataFrame
# convertir el diccionario en una lista recorriendo el diccionario items por items, nombrar las columnas que existiran en el df:
df_continentes = pd.DataFrame(list(diccionario_continentes.items()), columns=['Continente', 'Cantidad'])

In [53]:
df_continentes

Continente  Cantidad
0                           Africa        54
1                             Asia        49
2                           Europe        43
3  Latin America and the Caribbean        33
4                          Oceania        14
5                 Northern America         2

### Ejercicio 21:
- Repite los últimos dos ejercicios tomando en cuenta que america latina y america del norte son un solo continente.

In [54]:
#Vamos a seleccionar las filas que tengan en la columna "continente" latinoamerica y norteamerica
filas_acambiar = df[(df["Continente"] == "Latin America and the Caribbean") | (df["Continente"] == "Northern America")]


In [55]:
#Accedo a los indices de esas filas que hay que cambiar y ya puedo iterar porque es una lista
filas_acambiar.index
#lo podiamos haber puesto directamente arriba:

Int64Index([  6,   7,  12,  15,  18,  21,  24,  32,  35,  39,  42,  45,  52,
             53,  54,  56,  70,  71,  74,  75,  76,  86, 112, 125, 132, 134,
            135, 145, 146, 147, 165, 175, 185, 186, 189],
           dtype='int64')

In [56]:
filas_acambiar = df[(df["Continente"] == "Latin America and the Caribbean") | (df["Continente"] == "Northern America")].index
filas_acambiar

Int64Index([  6,   7,  12,  15,  18,  21,  24,  32,  35,  39,  42,  45,  52,
             53,  54,  56,  70,  71,  74,  75,  76,  86, 112, 125, 132, 134,
            135, 145, 146, 147, 165, 175, 185, 186, 189],
           dtype='int64')

In [57]:
for i in filas_acambiar:
    df.at[i, "Continente"] = "America" 

In [58]:
#crear un diccionario con la cantidad de paises de cada continente en la columna continente:
diccionario_continentes = df["Continente"].value_counts().to_dict()
diccionario_continentes

{'Africa': 54, 'Asia': 49, 'Europe': 43, 'America': 35, 'Oceania': 14}

In [59]:
#crear un df:
df_continentes = pd.DataFrame(list(diccionario_continentes.items()), columns=['Continente', 'Cantidad'])

In [60]:
df_continentes

Continente  Cantidad
0     Africa        54
1       Asia        49
2     Europe        43
3    America        35
4    Oceania        14

### Ejercicio 22:
- Crea un DataFrame por continente y en cada DataFrame apareceran los paises de dicho contiente.
- Muestra las 3 primeras filas de cada DataFrame.

In [61]:
#Crear una lista de todos los continentes:
diccionario_continentes = df["Continente"].value_counts().to_dict()
#accediendo a las llaves del diccionario:
lista_continentes = list(diccionario_continentes.keys())
lista_continentes

['Africa', 'Asia', 'Europe', 'America', 'Oceania']

In [62]:
for i in lista_continentes:
    print(i)

Africa
Asia
Europe
America
Oceania


In [63]:
for continente in lista_continentes:
    paises = df[df['Continente'] == continente]['Pais'].tolist()
    if continente == "Africa":
        df_africa = pd.DataFrame({"Pais": paises, "Continente" : continente})
    elif continente == "Asia":
        df_asia = pd.DataFrame({"Pais": paises, "Continente" : continente})
    elif continente == "Europe":
        df_europe = pd.DataFrame({"Pais": paises, "Continente" : continente})
    elif continente == "America":
        df_america = pd.DataFrame({"Pais": paises, "Continente" : continente})
    elif continente == "Oceania":
        df_oceania = pd.DataFrame({"Pais": paises, "Continente" : continente})

In [64]:
print(df_africa.shape, df_asia.shape, df_europe.shape, df_america.shape, df_oceania.shape)

(54, 2) (49, 2) (43, 2) (35, 2) (14, 2)


In [65]:
df_africa.head(3)

Pais Continente
0  Algeria     Africa
1   Angola     Africa
2    Benin     Africa

In [66]:
df_asia.head(3)

Pais Continente
0  Afghanistan       Asia
1      Armenia       Asia
2   Azerbaijan       Asia

In [67]:
df_europe.head(3)

Pais Continente
0  Albania     Europe
1  Andorra     Europe
2  Austria     Europe

In [68]:
df_america.head(3)

Pais Continente
0  Antigua and Barbuda    America
1            Argentina    America
2              Bahamas    America

In [69]:
df_oceania.head(3)

Pais Continente
0  American Samoa    Oceania
1       Australia    Oceania
2            Fiji    Oceania

### Ejercicio 23:
- Obtener los valores maximos de migración de cada año y crea un DataFrame con los años y el número de migración máximo de cada año.

|     |Año |Numero de Migrantes|
|-----|----|-------------------|
|**0**|1980|22045              |
|**1**|1981|24796              |
|**2**|1982|20620              |
|**3**|1983|10015              |
|**4**|1984|10170              |

In [76]:
#indice de la primera columna que nos interesa:
df.columns.get_loc(1980)

4

In [126]:
#indice de la ultima columna que nos interesa:
df.columns.get_loc(2013)

37

In [127]:
#Columnas interesantes: 1º LE SUMAMOS UN INDICE MAS
columnas = df.columns[4:38]

In [128]:
#valores maximos de cada columna interesante: 2º
valores_maximos = df[columnas].max()

In [129]:
#Creo un diccionario con los valores maximos: 3º
diccionario = valores_maximos.to_dict()

In [256]:
#Creo el dataframe: 4º
df_migracion = pd.DataFrame(list(diccionario.items()), columns= ["Año", "Nº Migrantes"])

In [257]:
df_migracion

Año  Nº Migrantes
0   1980         22045
1   1981         24796
2   1982         20620
3   1983         10015
4   1984         10170
5   1985          9564
6   1986          9470
7   1987         21337
8   1988         27359
9   1989         23795
10  1990         31668
11  1991         23380
12  1992         34123
13  1993         33720
14  1994         39231
15  1995         30145
16  1996         29322
17  1997         22965
18  1998         21049
19  1999         30069
20  2000         35529
21  2001         36434
22  2002         31961
23  2003         36439
24  2004         36619
25  2005         42584
26  2006         33848
27  2007         28742
28  2008         30037
29  2009         29622
30  2010         38617
31  2011         36765
32  2012         34315
33  2013         34129

In [132]:
#si quisiera que cada llave del diccionario fuese una columna y que los valores fueran una unica fila?
df_migracion2 = pd.DataFrame.from_dict(diccionario, orient='index').T

In [133]:
df_migracion2

1980   1981   1982   1983   1984  1985  1986   1987   1988   1989  ...  \
0  22045  24796  20620  10015  10170  9564  9470  21337  27359  23795  ...   

    2004   2005   2006   2007   2008   2009   2010   2011   2012   2013  
0  36619  42584  33848  28742  30037  29622  38617  36765  34315  34129  

[1 rows x 34 columns]

### Ejercicio 24:
- Ordenar por número mayor de migración a menor el DataFrame anterior.

In [258]:
df_migracion1 = df_migracion.sort_values(by= "Nº Migrantes", ascending=False)

In [259]:
df_migracion

Año  Nº Migrantes
0   1980         22045
1   1981         24796
2   1982         20620
3   1983         10015
4   1984         10170
5   1985          9564
6   1986          9470
7   1987         21337
8   1988         27359
9   1989         23795
10  1990         31668
11  1991         23380
12  1992         34123
13  1993         33720
14  1994         39231
15  1995         30145
16  1996         29322
17  1997         22965
18  1998         21049
19  1999         30069
20  2000         35529
21  2001         36434
22  2002         31961
23  2003         36439
24  2004         36619
25  2005         42584
26  2006         33848
27  2007         28742
28  2008         30037
29  2009         29622
30  2010         38617
31  2011         36765
32  2012         34315
33  2013         34129

### Ejercicio 25:
- Obtener el promedio y la mediana de cada año de migración y agregar al dataframe anterior.

In [260]:
df_migracion

Año  Nº Migrantes
0   1980         22045
1   1981         24796
2   1982         20620
3   1983         10015
4   1984         10170
5   1985          9564
6   1986          9470
7   1987         21337
8   1988         27359
9   1989         23795
10  1990         31668
11  1991         23380
12  1992         34123
13  1993         33720
14  1994         39231
15  1995         30145
16  1996         29322
17  1997         22965
18  1998         21049
19  1999         30069
20  2000         35529
21  2001         36434
22  2002         31961
23  2003         36439
24  2004         36619
25  2005         42584
26  2006         33848
27  2007         28742
28  2008         30037
29  2009         29622
30  2010         38617
31  2011         36765
32  2012         34315
33  2013         34129

In [261]:
df_migracion["Promedio"] = df_migracion["Nº Migrantes"].mean()
df_migracion["Mediana"] = df_migracion["Nº Migrantes"].median()

In [262]:
df_migracion #si nos fijamos el promedio lo hace de todos los valores de la columna n migrantes

Año  Nº Migrantes      Promedio  Mediana
0   1980         22045  28249.529412  30053.0
1   1981         24796  28249.529412  30053.0
2   1982         20620  28249.529412  30053.0
3   1983         10015  28249.529412  30053.0
4   1984         10170  28249.529412  30053.0
5   1985          9564  28249.529412  30053.0
6   1986          9470  28249.529412  30053.0
7   1987         21337  28249.529412  30053.0
8   1988         27359  28249.529412  30053.0
9   1989         23795  28249.529412  30053.0
10  1990         31668  28249.529412  30053.0
11  1991         23380  28249.529412  30053.0
12  1992         34123  28249.529412  30053.0
13  1993         33720  28249.529412  30053.0
14  1994         39231  28249.529412  30053.0
15  1995         30145  28249.529412  30053.0
16  1996         29322  28249.529412  30053.0
17  1997         22965  28249.529412  30053.0
18  1998         21049  28249.529412  30053.0
19  1999         30069  28249.529412  30053.0
20  2000         35529  28249.529412  30053.0
21  2001         36434  28249.529412  30053.0
22  2002         31961  28249.529412  30053.0
23  2003         36439  28249.529412  30053.0
24  2004         36619  28249.529412  30053.0
25  2005         42584  28249.529412  30053.0
26  2006         33848  28249.529412  30053.0
27  2007         28742  28249.529412  30053.0
28  2008         30037  28249.529412  30053.0
29  2009         29622  28249.529412  30053.0
30  2010         38617  28249.529412  30053.0
31  2011         36765  28249.529412  30053.0
32  2012         34315  28249.529412  30053.0
33  2013         34129  28249.529412  30053.0

In [137]:
#el promedio que querian era del df original... ponemos 2014, porque las filas y columnas empiezan en 0
df_migracion["Promedio Real"] = df[range(1980, 2014)].mean().to_list()

df_migracion["Mediana Real"] = df[range(1980, 2014)].median().to_list()

In [138]:
#Borramos las columnas con datos falsos que creamos anteriormente:
df_migracion.drop(["Promedio", "Mediana"], axis = 1, inplace = True)

In [139]:
df_migracion

Año  Nº Migrantes  Promedio Real  Mediana Real
0   1980         22045     508.394872          13.0
1   1981         24796     566.989744          10.0
2   1982         20620     534.723077          11.0
3   1983         10015     387.435897          12.0
4   1984         10170     376.497436          13.0
5   1985          9564     358.861538          17.0
6   1986          9470     441.271795          18.0
7   1987         21337     691.133333          26.0
8   1988         27359     714.389744          34.0
9   1989         23795     843.241026          44.0
10  1990         31668     964.379487          38.0
11  1991         23380    1064.148718          51.0
12  1992         34123    1136.856410          74.0
13  1993         33720    1138.712821          85.0
14  1994         39231     993.153846          76.0
15  1995         30145     962.625641          91.0
16  1996         29322    1026.076923         118.0
17  1997         22965     989.153846         114.0
18  1998         21049     824.241026         106.0
19  1999         30069     922.143590         116.0
20  2000         35529    1111.343590         138.0
21  2001         36434    1244.323077         169.0
22  2002         31961    1144.158974         165.0
23  2003         36439    1114.343590         161.0
24  2004         36619    1190.169231         191.0
25  2005         42584    1320.292308         210.0
26  2006         33848    1266.958974         218.0
27  2007         28742    1191.820513         198.0
28  2008         30037    1246.394872         205.0
29  2009         29622    1275.733333         214.0
30  2010         38617    1420.287179         211.0
31  2011         36765    1262.533333         179.0
32  2012         34315    1313.958974         233.0
33  2013         34129    1320.702564         213.0

In [ ]:
# Otra forma?
# df.loc[:, range(1980, 2014)].mean(axis=1).tolist()

### Ejercicio 26:
- Encontrar el año en que hubo más migración por continente y crear un dataframe con nombre de continente, el año de máxima migración y cantidad de migración.

|	  |Continente|Año |Migraciones|
|-----|----------|----|-----------|
|**0**|Asia	     |2005|42584      |
|**1**|Europe	 |1994|39231      |
|**2**|Africa	 |2010|6242       |
|**3**|Oceania	 |1992|1782       |
|**4**|America	 |2008|10190      |

In [ ]:
#nombres de continentes:
df["Continente"].unique()

In [168]:
#listas vacias:
continentes = list()
años = list()
migraciones = list()

#recorremos los continentes:
for cont in df["Continente"].unique():
    filas_porcontinente = df[df["Continente"] == cont]
    #filtra el DataFrame df para obtener solo las filas donde el valor en la columna "Continente" coincide con el valor actual de cont
    filas_desde1980a2013 = filas_porcontinente[range(1980,2014)]
    # selecciona las columnas desde 1980 hasta 2013
    max_columnas = filas_desde1980a2013.max()
    #Calcula el valor máximo para cada columna seleccionada (es decir, para cada año).
    max_columnas_ordenadas = max_columnas.sort_values()
    #Ordena los valores de máximo a mínimo. Esto significa que los años estarán ordenados desde el año con el máximo valor hasta el mínimo.
    año_maximo = max_columnas_ordenadas.tail(1)
    #el año con el valor maximo está en el ultimo valor
    nombre_año = año_maximo.index[0]
    cantidad_migracion = año_maximo.values[0]

    años.append(nombre_año)
    migraciones.append(cantidad_migracion)
    continentes.append(cont)
años, migraciones, continentes



([2005, 1994, 2010, 1992, 2008],
 [42584, 39231, 6242, 1782, 10190],
 ['Asia', 'Europe', 'Africa', 'Oceania', 'America'])

In [170]:
df_migracion = pd.DataFrame()

In [171]:
df_migracion["Continente"] = continentes
df_migracion["Año"] = años
df_migracion["Migraciones"] = migraciones

In [172]:
df_migracion

Continente   Año  Migraciones
0       Asia  2005        42584
1     Europe  1994        39231
2     Africa  2010         6242
3    Oceania  1992         1782
4    America  2008        10190

### Ejercicio 27:
- Encontrar los 10 paises con más migración entre el 2008 y 2010.

|       |Pais                                             |2008 |2009 |	2010|migracion_2010|
|-------|-------------------------------------------------|-----|-----|-----|--------------|
|**136**|Philippines                                      |24887|28573|38617|         92077|
|**79** |India                                            |28261|29456|34235|         91952|
|**36** |China                                            |30037|29622|30391|         90050|
|**185**|United States of America                         |10190|8995 |8142 |         27327|
|**183**|United Kingdom of Great Britain and Northern I...|8979 |8876 |8724 |         26579|
|**130**|Pakistan                                         |8994 |7217 |6811 |         23022|
|**81** |Iran (Islamic Republic of)                       |6475 |6580 |7477 |         20532|
|**140**|Republic of Korea                                |7294 |5874 |5537 |         18705|
|**116**|Morocco                                          |4226 |5532 |6242 |         16000|
|**39** |Colombia                                         |5452 |4652 |5218 |         15322|

In [177]:
#creamos una columna nueva donde aparezca la suma de las migraciones entre 2008 y 2010

df["Total_migracion_2008_2010"] = df[range(2008, 2011)].sum(axis = 1)

#ordenamos de mayor a menor el df, según los valores de la nueva columna, es decir, el total de migracion enttre 2008 y 2010

df.sort_values("Total_migracion_2008_2010", ascending = False, inplace = True)

#Mostramos del df solo las columnas que queremos y las 10 primeras filas:
df[["Pais", 2008, 2009, 2010, "Total_migracion_2008_2010"]].head(10)


Pais   2008   2009   2010  \
136                                        Philippines  24887  28573  38617   
79                                               India  28261  29456  34235   
36                                               China  30037  29622  30391   
185                           United States of America  10190   8995   8142   
183  United Kingdom of Great Britain and Northern I...   8979   8876   8724   
130                                           Pakistan   8994   7217   6811   
81                          Iran (Islamic Republic of)   6475   6580   7477   
140                                  Republic of Korea   7294   5874   5537   
116                                            Morocco   4226   5532   6242   
39                                            Colombia   5452   4652   5218   

     Total_migracion_2008_2010  
136                      92077  
79                       91952  
36                       90050  
185                      27327  
183                      26579  
130                      23022  
81                       20532  
140                      18705  
116                      16000  
39                       15322

### Ejercicio 28:
- Agregar una columna al DataFrame inicial con el número total de migración de 1980 al 2013 por pais.

In [178]:
df["Total_migracion"] = df[range(1980, 2014)].sum(axis = 1)

In [179]:
df

Pais Continente  \
136                                        Philippines       Asia   
79                                               India       Asia   
36                                               China       Asia   
185                           United States of America    America   
183  United Kingdom of Great Britain and Northern I...     Europe   
..                                                 ...        ...   
109                                   Marshall Islands    Oceania   
4                                              Andorra     Europe   
179                                             Tuvalu    Oceania   
131                                              Palau    Oceania   
32                                              Canada    America   

       Nombre de region      Tipo de region   1980   1981   1982   1983  \
136  South-Eastern Asia  Developing regions   6051   5921   5249   4562   
79        Southern Asia  Developing regions   8880   8670   8147   7338   
36         Eastern Asia  Developing regions   5123   6682   3308   1863   
185    Northern America   Developed regions   9378  10030   9074   7100   
183     Northern Europe   Developed regions  22045  24796  20620  10015   
..                  ...                 ...    ...    ...    ...    ...   
109          Micronesia  Developing regions      0      0      0      0   
4       Southern Europe   Developed regions      0      0      0      0   
179           Polynesia  Developing regions      0      1      0      0   
131          Micronesia  Developing regions      0      0      0      0   
32     Northern America   Developed regions      0      0      0      0   

      1984  1985  ...   2006   2007   2008   2009   2010   2011   2012   2013  \
136   3801  3150  ...  18400  19837  24887  28573  38617  36765  34315  29544   
79    5704  4211  ...  33848  28742  28261  29456  34235  27509  30933  33087   
36    1527  1816  ...  33518  27642  30037  29622  30391  28502  33024  34129   
185   6661  6543  ...   9613   9463  10190   8995   8142   7676   7891   8501   
183  10170  9564  ...   7140   8216   8979   8876   8724   6204   6195   5827   
..     ...   ...  ...    ...    ...    ...    ...    ...    ...    ...    ...   
109      0     0  ...      0      2      0      0      0      0      0      0   
4        0     0  ...      1      1      0      0      0      0      1      1   
179      1     0  ...      0      1      0      0      0      1      0      0   
131      0     0  ...      0      1      0      0      0      0      0      0   
32       0     0  ...      0      0      0      0      0      1      1      2   

     Total_migracion_2008_2010  Total_migracion  
136                      92077           511391  
79                       91952           691904  
36                       90050           659962  
185                      27327           241122  
183                      26579           551500  
..                         ...              ...  
109                          0                2  
4                            0               15  
179                          0                8  
131                          0                1  
32                           0               20  

[195 rows x 40 columns]

### Ejercicio 29:
- Encuentra el país de cada continente con más migración y crea un DataFrame con estos paises.

In [214]:
paises_mayormigracion = pd.DataFrame()
for cont in df["Continente"].unique():
    #ordenamos de mayor a menor por los valores de la columna "total_migracion" agrupando por continentes y cogemos la primera fila que aparezca:
    df[df["Continente"] == cont].sort_values("Total_migracion", ascending = False).head(1)
    #agregamos al dataframe vacio:
    paises_mayormigracion = paises_mayormigracion.append(df[df["Continente"] == cont].sort_values("Total_migracion", ascending = False).head(1))

C:\Users\Ane Zubieta\AppData\Local\Temp\ipykernel_13652\3048982608.py:6: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  paises_mayormigracion = paises_mayormigracion.append(df[df["Continente"] == cont].sort_values("Total_migracion", ascending = False).head(1))
C:\Users\Ane Zubieta\AppData\Local\Temp\ipykernel_13652\3048982608.py:6: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  paises_mayormigracion = paises_mayormigracion.append(df[df["Continente"] == cont].sort_values("Total_migracion", ascending = False).head(1))
C:\Users\Ane Zubieta\AppData\Local\Temp\ipykernel_13652\3048982608.py:6: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  paises_mayormigracion = paises_mayormigracion.append(df[df["Continente"] == cont].sort_values(

In [215]:
paises_mayormigracion

Pais Continente  \
79                                               India       Asia   
185                           United States of America    America   
183  United Kingdom of Great Britain and Northern I...     Europe   
55                                               Egypt     Africa   
9                                            Australia    Oceania   

              Nombre de region      Tipo de region   1980   1981   1982  \
79               Southern Asia  Developing regions   8880   8670   8147   
185           Northern America   Developed regions   9378  10030   9074   
183            Northern Europe   Developed regions  22045  24796  20620   
55             Northern Africa  Developing regions    612    660    755   
9    Australia and New Zealand   Developed regions    702    639    484   

      1983   1984  1985  ...   2006   2007   2008   2009   2010   2011   2012  \
79    7338   5704  4211  ...  33848  28742  28261  29456  34235  27509  30933   
185   7100   6661  6543  ...   9613   9463  10190   8995   8142   7676   7891   
183  10015  10170  9564  ...   7140   8216   8979   8876   8724   6204   6195   
55     455    447   348  ...   2190   2356   3347   3496   5982   4663   5555   
9      317    317   319  ...    875   1033   1018   1018    933    851    982   

      2013  Total_migracion_2008_2010  Total_migracion  
79   33087                      91952           691904  
185   8501                      27327           241122  
183   5827                      26579           551500  
55    4165                      12825            72745  
9     1121                       2969            23829  

[5 rows x 40 columns]

In [216]:
#ELIMINAR VARIAS COLUMNAS A LA VEZ POR SUS INDICES:
paises_mayormigracion.drop(paises_mayormigracion.columns[2:39], axis = 1, inplace = True)

In [228]:
paises_mayormigracion.reset_index()


index                                               Pais Continente  \
0     79                                              India       Asia   
1    185                           United States of America    America   
2    183  United Kingdom of Great Britain and Northern I...     Europe   
3     55                                              Egypt     Africa   
4      9                                          Australia    Oceania   

   Total_migracion  
0           691904  
1           241122  
2           551500  
3            72745  
4            23829

### Ejercicio 30:
- Encontrar los 3 paises con más migración de cada continente y agregarlos en un DataFrame.

|       |Continente|Pais                                             |Total Migracion|
|-------|----------|-------------------------------------------------|---------------|
|**79**	|Asia	   |India                                            |691904         |
|**36**	|Asia	   |China                                            |659962         |
|**136**|Asia	   |Philippines                                      |511391         |
|**183**|Europa	   |United Kingdom of Great Britain and Northern I...|551500         |
|**137**|Europa	   |Poland                                           |139241         |
|**63**	|Europa	   |France                                           |109091         |
|**55**	|Africa	   |Egypt                                            |72745          |
|**116**|Africa	   |Morocco                                          |72267          |
|**2**	|Africa	   |Algeria                                          |69439          |
|**9**	|Oceania   |Australia                                        |23829          |
|**61**	|Oceania   |Fiji                                             |20609          |
|**124**|Oceania   |New Zealand                                      |10386          |
|**185**|America   |United States of America                         |241122         |
|**86**	|America   |Jamaica                                          |106431         |
|**75**	|America   |Haiti                                            |84813          |

In [197]:
trespaises_mayormigracion = pd.DataFrame()
for cont in df["Continente"].unique():
    #ordenamos de mayor a menor por los valores de la columna "total_migracion" agrupando por continentes y cogemos las 3 primeras filas que aparezca:
    df[df["Continente"] == cont].sort_values("Total_migracion", ascending = False).head(3)
    #agregamos al dataframe vacio:
    trespaises_mayormigracion = trespaises_mayormigracion.append(df[df["Continente"] == cont].sort_values("Total_migracion", ascending = False).head(3))

C:\Users\Ane Zubieta\AppData\Local\Temp\ipykernel_13652\1275118875.py:6: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  trespaises_mayormigracion = trespaises_mayormigracion.append(df[df["Continente"] == cont].sort_values("Total_migracion", ascending = False).head(3))
C:\Users\Ane Zubieta\AppData\Local\Temp\ipykernel_13652\1275118875.py:6: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  trespaises_mayormigracion = trespaises_mayormigracion.append(df[df["Continente"] == cont].sort_values("Total_migracion", ascending = False).head(3))
C:\Users\Ane Zubieta\AppData\Local\Temp\ipykernel_13652\1275118875.py:6: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  trespaises_mayormigracion = trespaises_mayormigracion.append(df[df["Continente

In [210]:
trespaises_mayormigracion = trespaises_mayormigracion[["Continente", "Pais", "Total_migracion"]]

In [226]:
trespaises_mayormigracion.reset_index()


index Continente                                               Pais  \
0      79       Asia                                              India   
1      36       Asia                                              China   
2     136       Asia                                        Philippines   
3     185    America                           United States of America   
4      86    America                                            Jamaica   
5      75    America                                              Haiti   
6     183     Europe  United Kingdom of Great Britain and Northern I...   
7     137     Europe                                             Poland   
8      63     Europe                                             France   
9      55     Africa                                              Egypt   
10    116     Africa                                            Morocco   
11      2     Africa                                            Algeria   
12      9    Oceania                                          Australia   
13     61    Oceania                                               Fiji   
14    124    Oceania                                        New Zealand   

    Total_migracion  
0            691904  
1            659962  
2            511391  
3            241122  
4            106431  
5             84813  
6            551500  
7            139241  
8            109091  
9             72745  
10            72267  
11            69439  
12            23829  
13            20609  
14            10386

### Ejercicio 31:
- Cambiar los nombres de cada continente a español (utilizando un bucle for).

In [238]:
paises_mayormigracion["Continente"] = paises_mayormigracion["Continente"].replace("Europe", "Europa")
        

In [241]:
for cont in paises_mayormigracion["Continente"]:
    if cont == "America":
        paises_mayormigracion["Continente"] = paises_mayormigracion["Continente"].replace("America", "América")
    elif cont == "Africa":
        paises_mayormigracion["Continente"] = paises_mayormigracion["Continente"].replace("Africa", "África")
    elif cont == "Oceania":
        paises_mayormigracion["Continente"] = paises_mayormigracion["Continente"].replace("Oceania", "Oceanía")

In [243]:
paises_mayormigracion.reset_index()

index                                               Pais Continente  \
0     79                                              India       Asia   
1    185                           United States of America    América   
2    183  United Kingdom of Great Britain and Northern I...     Europa   
3     55                                              Egypt     África   
4      9                                          Australia    Oceanía   

   Total_migracion  
0           691904  
1           241122  
2           551500  
3            72745  
4            23829

#### Extra: Obtener un DataFrame que te parezca relevante a partir de la aplicación de filtros. ¡Se creativo!

In [245]:
df.reset_index()

index                                               Pais Continente  \
0      136                                        Philippines       Asia   
1       79                                              India       Asia   
2       36                                              China       Asia   
3      185                           United States of America    America   
4      183  United Kingdom of Great Britain and Northern I...     Europe   
..     ...                                                ...        ...   
190    109                                   Marshall Islands    Oceania   
191      4                                            Andorra     Europe   
192    179                                             Tuvalu    Oceania   
193    131                                              Palau    Oceania   
194     32                                             Canada    America   

       Nombre de region      Tipo de region   1980   1981   1982   1983  \
0    South-Eastern Asia  Developing regions   6051   5921   5249   4562   
1         Southern Asia  Developing regions   8880   8670   8147   7338   
2          Eastern Asia  Developing regions   5123   6682   3308   1863   
3      Northern America   Developed regions   9378  10030   9074   7100   
4       Northern Europe   Developed regions  22045  24796  20620  10015   
..                  ...                 ...    ...    ...    ...    ...   
190          Micronesia  Developing regions      0      0      0      0   
191     Southern Europe   Developed regions      0      0      0      0   
192           Polynesia  Developing regions      0      1      0      0   
193          Micronesia  Developing regions      0      0      0      0   
194    Northern America   Developed regions      0      0      0      0   

      1984  ...   2006   2007   2008   2009   2010   2011   2012   2013  \
0     3801  ...  18400  19837  24887  28573  38617  36765  34315  29544   
1     5704  ...  33848  28742  28261  29456  34235  27509  30933  33087   
2     1527  ...  33518  27642  30037  29622  30391  28502  33024  34129   
3     6661  ...   9613   9463  10190   8995   8142   7676   7891   8501   
4    10170  ...   7140   8216   8979   8876   8724   6204   6195   5827   
..     ...  ...    ...    ...    ...    ...    ...    ...    ...    ...   
190      0  ...      0      2      0      0      0      0      0      0   
191      0  ...      1      1      0      0      0      0      1      1   
192      1  ...      0      1      0      0      0      1      0      0   
193      0  ...      0      1      0      0      0      0      0      0   
194      0  ...      0      0      0      0      0      1      1      2   

     Total_migracion_2008_2010  Total_migracion  
0                        92077           511391  
1                        91952           691904  
2                        90050           659962  
3                        27327           241122  
4                        26579           551500  
..                         ...              ...  
190                          0                2  
191                          0               15  
192                          0                8  
193                          0                1  
194                          0               20  

[195 rows x 41 columns]